lib/fields.ts

```diff
@@ -0,0 +1,70 @@
+// Helper functions for creating fields
+
+import { FieldDefinition, ParquetType, SchemaDefinition } from "./declare";
+
+export function createStringField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'UTF8' };
+}
+
+export function createBooleanField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'BOOLEAN' };
+}
+
+export function createIntField(size: 32 | 64, optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: `INT${size}` };
+}
+
+export function createFloatField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'FLOAT' };
+}
+
+export function createDoubleField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'DOUBLE' };
+}
+
+export function createDecimalField(precision: number, optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, precision, optional, type: 'FLOAT' };
+}
+
+export function createTimestampField(optional = true, fieldOptions: FieldDefinition = {}): FieldDefinition {
+    return { ...fieldOptions, optional, type: 'TIMESTAMP_MILLIS' };
+}
+
+export function createStructField(fields: SchemaDefinition, optional = true): FieldDefinition {
+    return {
+        optional,
+        fields,
+    }
+}
+
+export function createStructListField(fields: SchemaDefinition, optional = true): FieldDefinition {
+    return {
+        type: 'LIST',
+        repeated: true,
+        optional,
+        fields: {
+            list: {
+                fields,
+            },
+        },
+    }
+}
+
+export function createListField(type: ParquetType, optionalElement = false, optionalListField = true, elementOptions: FieldDefinition = {}): FieldDefinition {
+    return {
```

Comment:
same test case as for `createStructListField`.

The error message is the following:
```
Invalid list type repeated group array_field (LIST) {
  required group list {
    required binary element (STRING);
  }
}

```

If I change the return value to:

```js
  return {
    type: "LIST",

    optional: optionalListField,
    fields: {
      list: {
        repeated: true,
        fields: {
          element: {
            ...elementOptions,
            type,
            optional: optionalElement,
          },
        },
      },
    },
  }
```

The error doesn't occur.

Reply:
Ok. Took a bit to get through it all, but pushed up changes and taking your example into a test as well.

Oddly enough [Tad](https://github.com/antonycourtney/tad) opened the "bad" files without error, but what really helped lock it in was https://github.com/manojkarthick/pqrs

Reply:
Didn't know both tools - thanks for sharing